drivers/kw2xrf/kw2xrf_tx.c

```diff
@@ -0,0 +1,321 @@
+    switch (kind) {
+        case NETDEV_802154_PKT_KIND_BEACON:
+            hdr[0] = 0x00;
+            break;
+
+        case NETDEV_802154_PKT_KIND_DATA:
+            hdr[0] = 0x01;
+            break;
+
+        case NETDEV_802154_PKT_KIND_ACK:
+            hdr[0] = 0x02;
+            break;
+
+        default:
+            return NETDEV_802154_TX_STATUS_INVALID_PARAM;
+    }
+
+    if (wants_ack) {
+        hdr[0] |= 0x20;
+    }
+
+    wait_for_ack = wants_ack;
+
+    uint16_t src_pan = kw2xrf_get_pan();
+    bool compress_pan = false;
+
+    if (use_long_addr) {
+        hdr[1] = 0xcc;
+    }
+    else {
+        hdr[1] = 0x88;
+
+        /* short address mode, use PAN ID compression
+            for intra-PAN communication */
+        if (dest->pan.id == src_pan) {
+            compress_pan = true;
+            hdr[0] |= 0x40;
+        }
+    }
+
+    /* sequence number */
+    hdr[2] = sequence_nr++;
+
+    /* variable-length fields */
+    int idx = 3;
+
+    if (use_long_addr) {
+        /* dest long addr */
+        hdr[idx++] = (uint8_t)(dest->long_addr & 0xFF);
+        hdr[idx++] = (uint8_t)(dest->long_addr >> 8);
+        hdr[idx++] = (uint8_t)(dest->long_addr >> 16);
+        hdr[idx++] = (uint8_t)(dest->long_addr >> 24);
+        hdr[idx++] = (uint8_t)(dest->long_addr >> 32);
+        hdr[idx++] = (uint8_t)(dest->long_addr >> 40);
+        hdr[idx++] = (uint8_t)(dest->long_addr >> 48);
+        hdr[idx++] = (uint8_t)(dest->long_addr >> 56);
+        /* src long addr */
+        uint64_t src_long_addr = kw2xrf_get_address_long();
+        hdr[idx++] = (uint8_t)(src_long_addr & 0xFF);
+        hdr[idx++] = (uint8_t)(src_long_addr >> 8);
+        hdr[idx++] = (uint8_t)(src_long_addr >> 16);
+        hdr[idx++] = (uint8_t)(src_long_addr >> 24);
+        hdr[idx++] = (uint8_t)(src_long_addr >> 32);
+        hdr[idx++] = (uint8_t)(src_long_addr >> 40);
+        hdr[idx++] = (uint8_t)(src_long_addr >> 48);
+        hdr[idx++] = (uint8_t)(src_long_addr >> 56);
+    }
+    else {
+        /* dest PAN ID */
+        hdr[idx++] = (uint8_t)(dest->pan.id & 0xFF);
+        hdr[idx++] = (uint8_t)(dest->pan.id >> 8);
+        /* dest short addr */
+        hdr[idx++] = (uint8_t)(dest->pan.addr & 0xFF);
+        hdr[idx++] = (uint8_t)(dest->pan.addr >> 8);
+
+        /* src PAN ID */
+        if (!compress_pan) {
+            hdr[idx++] = (uint8_t)(src_pan & 0xFF);
+            hdr[idx++] = (uint8_t)(src_pan >> 8);
+        }
+
+        /* src short addr */
+        uint16_t src_addr = kw2xrf_get_address();
+        hdr[idx++] = (uint8_t)(src_addr & 0xFF);
+        hdr[idx++] = (uint8_t)(src_addr >> 8);
+    }
+
+    /* total frame size */
+    uint8_t size = idx + len + 2;
+
+    if (size > MKW2XDRF_MAX_PKT_LENGTH) {
+        return NETDEV_802154_TX_STATUS_PACKET_TOO_LONG;
+    }
+
+    pkt[0] = size;
+
+    /* generate pkt */
+    for (uint8_t i = 0; i < len; i++) {
+        hdr[i + idx] = ((uint8_t *)buf)[i];
```

Comment:
At this point it might be more efficient to call the write_fifo-function only once and write the complete packet at once via SPI.